opensoundscape/audio.py

```diff
@@ -266,79 +286,71 @@ def split(self, clip_duration, clip_overlap=0, final_clip=None):
+                By default, ignores final clip entirely.
+                Possible options (any other input will ignore the final clip entirely),
+                    - "remainder":  Include the remainder of the Audio (clip will not have clip_duration length)
+                    - "full":       Increase the overlap to yield a clip with clip_duration length
+                    - "extend":     Similar to remainder but extend (repeat) the clip to reach clip_duration length
         Returns:
             A list of dictionaries with keys: ["audio", "begin_time", "end_time"]
         """
 
         duration = self.duration()
-        if clip_duration > duration:
-            if final_clip == "remainder":
-                return_clip = Audio(
-                    self.samples,
-                    self.sample_rate,
-                    resample_type=self.resample_type,
-                    max_duration=self.max_duration,
-                )
-                return [
-                    {
-                        "clip": return_clip,
-                        "clip_duration": return_clip.duration(),
-                        "begin_time": 0,
-                        "end_time": duration,
-                    }
-                ]
-            elif final_clip in ["full", "extend"]:
-                return_clip = self.extend(clip_duration)
-                return [
-                    {
-                        "clip": return_clip,
-                        "clip_duration": return_clip.duration(),
-                        "begin_time": 0,
-                        "end_time": duration,
-                    }
-                ]
-            else:
-                warnings.warn(
-                    f"Given Audio object with duration of `{duration}` seconds and `clip_duration={clip_duration}` but `final_clip={final_clip}` produces no clips. Returning empty list."
-                )
-                return []
 
-        num_clips = ceil((duration - clip_overlap) / (clip_duration - clip_overlap))
-        to_return = [None] * num_clips
-        for idx in range(num_clips):
-            if idx == num_clips - 1:
+        # Lists of start and end times for clips
+        increment = clip_duration - clip_overlap
+        starts = np.arange(0, duration, increment)
+        ends = starts + clip_duration
+
+        # Remove final_clip if needed
+        if final_clip not in ["remainder", "full", "extend"]:
+            # Throw away any clips with end times beyond the duration
+            keeps = ends <= duration
+            ends = ends[keeps]
+            starts = starts[keeps]
+
+        # Now we have the starts and ends
+        final_idx = len(ends) - 1
+        to_return = [None] * (final_idx + 1)
+        for idx, (start, end) in enumerate(zip(starts, ends)):
+            # By default
+            begin_time = start
+            end_time = end
+
+            # Change defaults to handle final clip
+            if idx >= final_idx:
```

Comment:
since this is a very critical feature - are you sure that it behaves the same way as the old code? Or if there are differences in how it handles corner cases, what are they?

Reply:
I am pretty confident this code behaves the same as the old code. It passes all the tests that the previous split code passed, plus a bunch of new splitting tests I added to `test_audio.py`. And vice versa, when I apply the new tests to the old splitting code, the old code passes all of the new tests except for a test that catches the corner case described below.

The one corner case that differs is what I thought was unexpected/undesirable behavior in the previous code: when using `final_clip="full"`, the old split code casts the start time of the final clip as an integer. This means that when making non-overlapping 5s clips from an Audio object that is, say, 8.2s long, the old code extracted the final clip at (3s, 8s) instead of (3.2s, 8.2s). If for some reason that's the desired behavior, we can change it back pretty easily by changing this line back to `begin_time = int(duration - clip_duration)`: https://github.com/kitzeslab/opensoundscape/blob/0316ba06ba535fba62d5429b7c62feb67f160941/opensoundscape/audio.py#L337